rust/src/client.rs

```diff
@@ -0,0 +1,166 @@
+use crate::service::Handle;
+use crate::coordinator::RoundParameters;
+use crate::participant::{Participant, Task};
+use crate::{CoordinatorPublicKey, SeedDict, SumDict, PetError, InitError};
+use sodiumoxide::crypto::box_;
+use std::sync::Arc;
+use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
+use tokio::time;
+use std::time::Duration;
+
+
+/// Client-side errors
+pub enum ClientError {
+    /// Error starting the underlying `Participant`
+    ParticipantInitErr(InitError),
+
+    /// Error from the underlying `Participant`
+    ParticipantErr(PetError),
+
+    /// Error deserialising service data
+    DeserialiseErr(bincode::Error),
+}
+
+/// A client of the federated learning service
+///
+/// `Client` is responsible for communicating with the service, deserialising
+/// its messages and delegating their processing to the underlying `Participant`.
+pub struct Client {
+    /// Handle to the federated learning `Service`
+    handle: Handle,
+
+    /// The underlying `Participant`
+    particip: Participant,
+
+    /// Interval to poll for service data
+    interval: time::Interval,
+
+    // TODO global model
+}
+
+impl Client {
+    /// Create a new `Client`
+    ///
+    /// `period`: time period at which to poll for service data, in seconds.
+    /// Returns `Ok(client)` if `Client` `client` initialised successfully
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    pub fn new(period: u64) -> Result<Self, ClientError> {
+        let (handle, _events) = Handle::new();
+        let particip = Participant::new()
+            .map_err(ClientError::ParticipantInitErr)?;
+        Ok(Self {
+            handle,
+            particip,
+            interval: time::interval(Duration::from_secs(period)),
+        })
+    }
+
+    /// Start the `Client` loop
+    ///
+    /// Returns `Err(err)` if `ClientError` `err` occurred
+    /// NOTE in the future this may iterate only a fixed number of times, at the
+    /// end of which this returns `Ok(())`
+    pub async fn start(&mut self) -> Result<(), ClientError> {
+        loop {
+            // any error that bubbles up will finish off the client
+            if let Err(e) = self.per_round().await {
```

Comment:
nice, I didn't know this could work, since the future returned by `pre_round` is not `'static`.

Reply:
This could be simplified a little:

```rust
loop {
    self.per_round().await?;
}
```